src/Data/Text.hs

```diff
@@ -1032,18 +1032,18 @@ foldr' f z t = S.foldl' (P.flip f) z (reverseStream t)
 
 -- | /O(n)/ Concatenate a list of 'Text's.
 concat :: [Text] -> Text
-concat ts = case ts' of
-              [] -> empty
-              [t] -> t
-              _ -> Text (A.run go) 0 len
+concat ts = case ts of
+    [] -> empty
+    [t] -> t
+    _ | len == 0 -> empty
+      | otherwise -> Text (A.run go) 0 len
   where
-    ts' = L.filter (not . null) ts
-    len = sumP "concat" $ L.map lengthWord8 ts'
+    len = sumP "concat" $ L.map lengthWord8 ts
     go :: ST s (A.MArray s)
     go = do
       arr <- A.new len
       let step i (Text a o l) = A.copyI l arr i a o >> return (i + l)
-      foldM step 0 ts' >> return arr
+      foldM step 0 ts >> return arr
```

Comment:
Note: I noticed that the index in this loop `i` stays boxed with `-O`, but not with `-O2`. Since `text` is compiled with `-O2`, it's not a problem, but let me know if you would like this changed to something that is unboxed with `-O`.

Reply:
As you say, I'm also curious to know what the optimizer at `-O` is missing but `-O2` for text seems a safe enough assumption.

Reply:
Yes I wouldn't be able to explain why this needs `O2` (`fspec-constr` is also enough).  
If you want me to tweak this, this is as small a change as `!_ <- foldM step 0 ts`. But I'm okay with leaving this as is if you are.